src/Stryker.Core/Stryker.Core/Mutants/MutationStore.cs

```diff
@@ -0,0 +1,125 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.Extensions.Logging;
+using Stryker.Core.Logging;
+
+namespace Stryker.Core.Mutants
+{
+    public enum MutationControl
+    {
+        Expression,
+        Statement,
+        Block
+    }
+
+    /// <summary>
+    /// Keeps track of mutations while parsing
+    /// </summary>
+    public class MutationStore
+    {
+        private static readonly ILogger _logger;
+        protected List<Mutant> _expressionMutants = new();
+        protected Stack<List<Mutant>> _statementMutants = new();
+        protected Stack<List<Mutant>> _blockMutants = new();
+
+        static MutationStore()
+        {
+            _logger = ApplicationLogging.LoggerFactory.CreateLogger<MutationStore>();
+        }
+
+        public bool HasBlockLevel => _blockMutants.Count > 0 && _blockMutants.Peek().Count > 0;
+
+        public void StoreMutations(IEnumerable<Mutant> proposedMutations, MutationControl control)
+        {
+            var mutations = proposedMutations.Where(m => m.ResultStatus == MutantStatus.NotRun);
+            if (!mutations.Any())
+            {
+                return;
+            }
+            switch (control)
+            {
+                case MutationControl.Expression:
+                    _expressionMutants.AddRange(mutations);
+                    break;
+                case MutationControl.Statement:
+                    if (_statementMutants.Count == 0)
+                    {
+                        // try again at block level
+                        StoreMutations(mutations, MutationControl.Block);
+                        return;
+                    }
+                    _statementMutants.Peek().AddRange(mutations);
+                    break;
+                case MutationControl.Block:
+                    if (_blockMutants.Count == 0)
+                    {
+                        // we have mutations that are going to be lost
+                        _logger.LogInformation($"{mutations.Count()} were generated but could not be injected as they cannot be controlled dynamically.");
```

Comment:
I don't think the user will understand what's happening here based on the message. Informational logging should imo be obvious or explicit about why the user should care about this. Does it cause inconsistency? Should the user care? Can the user do something about this?

Reply:
Agreed, changed to LogDebug